### ISSUE TITLE
Next.js 13: SWC Emotion Transform Plugin Breaks with root layout Server Component in `app/`

### ISSUE THREAD

**Issue by karlhorky** [vercel/next.js#41994]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```
    Operating System:
      Platform: linux
      Arch: x64
      Version: Ubuntu 20.04.0 LTS Thu Oct 27 2022 20:19:36 GMT+0200 (Central European Summer Time)
    Binaries:
      Node: 16.14.2
      npm: 7.17.0
      Yarn: 1.22.19
      pnpm: 7.13.6
    Relevant packages:
      next: 13.0.1-canary.0
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
```

### Describe the Bug

Using the `compiler: { emotion: true }` option is throwing an error about `React.createContext` not being a function:

StackBlitz: https://stackblitz.com/edit/vercel-next-js-nlknru?file=app%2Fpage.tsx,app%2Flayout.tsx,app%2FEmotionRootStyleRegistry.tsx,package.json

```
event - compiled client and server successfully in 59 ms (403 modules)
error - (sc_server)/node_modules/@emotion/react/dist/emotion-element-b63ca7c6.cjs.dev.js (20:47) @ eval
error - TypeError: React.createContext is not a function
    at eval (webpack-internal:///(sc_server)/./node_modules/@emotion/react/dist/emotion-element-b63ca7c6.cjs.dev.js:19:49)
    at (sc_server)/./node_modules/@emotion/react/dist/emotion-element-b63ca7c6.cjs.dev.js (/home/projects/vercel-next-js-mxnxa7/.next/server/app/page.js:501:1)
    at __webpack_require__ (/home/projects/vercel-next-js-mxnxa7/.next/server/webpack-runtime.js:33:43)
    at eval (webpack-internal:///(sc_server)/./node_modules/@emotion/react/jsx-dev-runtime/dist/emotion-react-jsx-dev-runtime.cjs.dev.js:7:22)
    at (sc_server)/./node_modules/@emotion/react/jsx-dev-runtime/dist/emotion-react-jsx-dev-runtime.cjs.dev.js (/home/projects/vercel-next-js-mxnxa7/.next/server/app/page.js:512:1)
    at __webpack_require__ (/home/projects/vercel-next-js-mxnxa7/.next/server/webpack-runtime.js:33:43)
    at eval (webpack-internal:///(sc_server)/./app/layout.tsx:5:88)
    at (sc_server)/./app/layout.tsx (/home/projects/vercel-next-js-mxnxa7/.next/server/app/page.js:403:1)
    at __webpack_require__ (/home/projects/vercel-next-js-mxnxa7/.next/server/webpack-runtime.js:33:43)
    at Object.layout (webpack-internal:///(sc_server)/./node_modules/next/dist/build/webpack/loaders/next-app-loader.js?name=app%2Fpage&appPaths=%2Fpage&pagePath=private-next-app-dir%2Fpage.tsx&appDir=%2Fhome%2Fprojects%2Fvercel-next-js-mxnxa7%2Fapp&pageExtensions=tsx&pageExtensions=ts&pageExtensions=jsx&pageExtensions=js&rootDir=%2Fhome%2Fprojects%2Fvercel-next-js-mxnxa7&isDev=true&tsconfigPath=tsconfig.json!:22:99) {
  type: 'TypeError',
  page: '/'
}
null
```

![Screenshot 2022-10-27 at 19 46 47](https://user-images.githubusercontent.com/1935696/198362773-4be2472a-0954-47ef-a588-75d5d7743639.png)

---

Making the root layout component into a client component does work, but this would be unfortunate:

StackBlitz: https://stackblitz.com/edit/vercel-next-js-rtrrxg?file=app%2Flayout.tsx,app%2Fpage.tsx,app%2FEmotionRootStyleRegistry.tsx,package.json

---

**Workaround:**

Disable the SWC Emotion transform plugin and use the `/** @jsxImportSource @emotion/react */` directive instead:

- https://github.com/emotion-js/emotion/issues/2928#issuecomment-1293885043

### Expected Behavior

It works to use the SWC Emotion transform plugin via `compiler: { emotion: true }` with a server component as the root layout

### Link to reproduction

https://stackblitz.com/edit/vercel-next-js-nlknru?file=app%2Fpage.tsx,app%2Flayout.tsx,app%2FEmotionRootStyleRegistry.tsx,package.json

### To Reproduce

1. Install `next@13.0.1-canary.0`, `@emotion/cache@11.10.5`, `@emotion/react@11.10.5`
2. Enable the `app/` dir and the SWC Emotion Transform plugin using `compiler: { emotion: true }, experimental: { appDir: true }` in `next.config.js`
3. Add the files as mentioned in @mitchellhamilton's post here: https://github.com/emotion-js/emotion/issues/2928#issuecomment-1293012737
4. Remove the `/** @jsxImportSource @emotion/react */` directive to rely on the SWC Emotion transform plugin to add `css` prop support
5. Start the server and observe the error

<sub>[NEXT-1368](https://linear.app/vercel/issue/NEXT-1368/nextjs-13-swc-emotion-transform-plugin-breaks-with-root-layout-server)</sub>

**Comment by r0manr0k0n** [comment]
In the same way, I experience the same problem with @mui, which uses emotion under the hood.

**Comment by AlejandroRodarte** [comment]
Reporting a similar problem with [nextui](https://github.com/nextui-org/nextui).

**Comment by talentlessguy** [comment]
Same with `@react-spring/web`

**Comment by rolanday** [comment]
Ditto. mui5, which depends on emotion breaks with ```TypeError: React.createContext is not a function``` when attempting to wrap app layout with theme context provider.

**Comment by Romanapf** [comment]
Same issue with framer-motion library

**Comment by karlhorky** [comment]
Maybe the people replying about different libraries here can also mention whether your reproduction uses SWC with `compiler: { emotion: true }` in `next.config.js` (or maybe the library configures this internally for you).

If you are not using this config option, then actually your comments are not related to the subject of this issue.

**Comment by dan-turner** [comment]
@karlhorky are you saying there's a correct way to make Next.JS v13 server components work with Emotion? Sorry new to this space

**Comment by rolanday** [comment]
> 
> Maybe the people replying about different libraries here can also mention whether your reproduction uses SWC with `compiler: { emotion: true }` in `next.config.js` (or maybe also look if your library configures this internally for you). And then edit your posts to include this information.
> 
> If you are not using this config option, then actually your comments are not related to the subject of this issue.

I am using a vanilla Next 13.0.0 config created using `npx create-next-app@latest --typescript`, and the only changes are 1) setting `experimental.appDir` to `true`, and 2) addition of `layout.tsx` with MUI v5 Box component added, plus a no-op page.tsx. Results in same exact error as op posted.

```
/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  swcMinify: true,
  experimental: {
    appDir: true,
  }
}

module.exports = nextConfig
```

```
- project root
   - app
      - layout.tsx
      - page.tsx

```

```
// layout.tsx
import { Box } from '@mui/material';

export default function RootLayout({
  children,
}: {
  children: React.ReactNode
}) {
  return (
    <html>
      <head></head>
      <Box sx={{ display: 'flex', flexDirection: 'column' }}>
        <body>{children}</body>
      </Box>
    </html>
  )
}
```

```
// page.tsx
export default function Page() {
  return <div></div>
}
```

**Comment by karlhorky** [comment]
That may be unrelated enough for a new issue (for MUI without this `compiler: { emotion: true }` config)

**Comment by mwskwong** [comment]
As far as I recall, the doc mentioned Emotion doesn't work with `app/` at all isn't it?

**Comment by codyslex1a** [comment]
I have the same problem with React-redux. When trying to wrap `{children}` with a `store` Provider:

```TypeError: React.createContext is not a function```

**Comment by dan-turner** [comment]
> @karlhorky are you saying there's a correct way to make Next.JS v13 server components work with Emotion? Sorry new to this space
> 
> Edit: just noticed the workaround mentioned above. Will give it a try

Doesn't seem to work without "use client"

**Comment by karlhorky** [comment]
> Doesn't seem to work without "use client"

There is a version that works with a root layout as a server component. The page and provider is still a client component. You can find this workaround link in the section marked **Workaround** in the description (the one mentioning `@jsxImportSource @emotion/react`).

**Comment by Rafcin** [comment]
@karlhorky To what description are you referring?

**Comment by creazy231** [comment]
Same problem using `react-select` https://github.com/jedwatson/react-select

**Comment by magnuen2k** [comment]
## Issues with different libraries workaround

> I have the same problem with React-redux. When trying to wrap `{children}` with a `store` Provider:
> 
> `TypeError: (0 , _react.createContext) is not a function`


> 

> Same issue with `framer-motion` library

## Workaround

My workaround was creating my own custom wrapper for the providers with "use client" like this:

### Custom Wrapper `Wrapper.tsx`

```jsx
'use client'

export const Wrapper = ({ children }: { children: React.ReactNode }) => {
  return (
    <Provider store={store}>
      {children}
    </Provider>
  );
};
```

### Then wrapping the children in `layout.tsx` like this:

```jsx
export default function RootLayout({
  children,
}: {
  children: React.ReactNode;
}) {
  return (
    <html lang="en">
      <body>
        <Wrapper>{children}</Wrapper>
      </body>
    </html>
  );
}
```

Not sure if this is optimal, but it works for now.

**Comment by tisonkun** [comment]
@magnuen2k I believe we should finally avoid `'use client'` since we shouldn't need a client module in this case.

**Comment by magnuen2k** [comment]
@tisonkun I think the issue here is that you cannot use context etc. without specifying that the component is a client component. And react-redux etc. does not specify that.

**Comment by iSuslov** [comment]
**Overall, is it valid to say that MUI does not work with Nextjs v13?**

**Comment by Rafcin** [comment]
@iSuslov I mean it works, but yea you can't leverage app dir etc. I think soon we'll start seeing solutions. People depend on MUI and people use CSS-in-JS solutions for a lot of existing projects so a fix will come.

**Comment by AChristoff** [comment]
Yep I'm stuck too. I can`t migrate to v13 because the project depends on MUI

**Comment by Daavidaviid** [comment]
So we cannot use **react-spring** currently with server component and **Nextjs v13** ?

**Comment by JustKira** [comment]
if anyone is using some sort of wrapper make sure to define "use client" in frist line of script of wrapper. anything that has useSomthing in it must be in script that start with "use client"

**Comment by antiproblemist** [comment]
Same issue exists when using antd with NextJS v13 appdir
The workaround I used was to declare `'use client'` on top of layout.js where antd components were imported
Still waiting for a proper fix

**Comment by Zubayer94** [comment]
> I have the same problem with React-redux. When trying to wrap `{children}` with a `store` Provider:
> 
> `TypeError: (0 , _react.createContext) is not a function`
> 
> ### Call Stack
> ```lisp
> eval
> 	(sc_server)/node_modules/react-redux/lib/components/Context.js (8:65)
> Object.(sc_server)/./node_modules/react-redux/lib/components/Context.js
> 	file:///Users/cody/Code/lafuong-web/.next/server/app/page.js (5035:1)
> __webpack_require__
> 	file:///Users/cody/Code/lafuong-web/.next/server/webpack-runtime.js (33:43)
> eval
> 	(sc_server)/node_modules/react-redux/lib/hooks/useReduxContext.js (8:15)
> Object.(sc_server)/./node_modules/react-redux/lib/hooks/useReduxContext.js
> 	file:///Users/cody/Code/lafuong-web/.next/server/app/page.js (5167:1)
> __webpack_require__
> 	file:///Users/cody/Code/lafuong-web/.next/server/webpack-runtime.js (33:43)
> eval
> 	(sc_server)/node_modules/react-redux/lib/hooks/useSelector.js (9:23)
> Object.(sc_server)/./node_modules/react-redux/lib/hooks/useSelector.js
> 	file:///Users/cody/Code/lafuong-web/.next/server/app/page.js (5178:1)
> __webpack_require__
> 	file:///Users/cody/Code/lafuong-web/.next/server/webpack-runtime.js (33:43)
> eval
> 	(sc_server)/node_modules/react-redux/lib/index.js (22:19)```
> ```

@

> codyslex1a
did you find any solution ?

**Comment by Zubayer94** [comment]
> I have the same problem with React-redux. When trying to wrap `{children}` with a `store` Provider:
> 
> `TypeError: (0 , _react.createContext) is not a function`
> 
> ### Call Stack
> ```lisp
> eval
> 	(sc_server)/node_modules/react-redux/lib/components/Context.js (8:65)
> Object.(sc_server)/./node_modules/react-redux/lib/components/Context.js
> 	file:///Users/cody/Code/lafuong-web/.next/server/app/page.js (5035:1)
> __webpack_require__
> 	file:///Users/cody/Code/lafuong-web/.next/server/webpack-runtime.js (33:43)
> eval
> 	(sc_server)/node_modules/react-redux/lib/hooks/useReduxContext.js (8:15)
> Object.(sc_server)/./node_modules/react-redux/lib/hooks/useReduxContext.js
> 	file:///Users/cody/Code/lafuong-web/.next/server/app/page.js (5167:1)
> __webpack_require__
> 	file:///Users/cody/Code/lafuong-web/.next/server/webpack-runtime.js (33:43)
> eval
> 	(sc_server)/node_modules/react-redux/lib/hooks/useSelector.js (9:23)
> Object.(sc_server)/./node_modules/react-redux/lib/hooks/useSelector.js
> 	file:///Users/cody/Code/lafuong-web/.next/server/app/page.js (5178:1)
> __webpack_require__
> 	file:///Users/cody/Code/lafuong-web/.next/server/webpack-runtime.js (33:43)
> eval
> 	(sc_server)/node_modules/react-redux/lib/index.js (22:19)```
> ```

@codyslex1a did you find any solution? I faced the same problem here....

**Comment by antikvarBE** [comment]
> So we cannot use **react-spring** currently with server component and **Nextjs v13** ?

I know I can't. 'use client' workaround is not doing it for me even if I truffle all of my components with it.

**Comment by Daavidaviid** [comment]
> > So we cannot use **react-spring** currently with server component and **Nextjs v13** ?
> 
> 
> 
> I know I can't. 'use client' workaround is not doing it for me even if I truffle all of my components with it.

Now that I've been using nextjs13appDir for a little while I realize how stupid my previous comment was 😅

'use client' is working for me.
Just try using it as deep as you can in the tree of components.

Only where it's necessary to not lose the optimization of Server component

**Comment by rtrembecky** [comment]
guys. @karlhorky tried to explain multiple times - this issue is about the compiler option in `next.config.js` AFTER you work around the basic emotion or MUI setup using `"use client"` directives for `CacheProvider` and all the components that will use emotion. If you comment out the `emotion: true` line in `next.config.js` in his StackBlitz (https://stackblitz.com/edit/vercel-next-js-nlknru?file=next.config.js), the app doesn't throw, but also it doesn't work (the rendered text isn't green). You currently need another workaround because of this, as shown in the second @karlhorky's StackBlitz (https://stackblitz.com/edit/vercel-next-js-ajvkxp?file=app/page.tsx) - put `/** @jsxImportSource @emotion/react */` on top of `page.tsx`.

**Comment by Systemcluster** [comment]
Adding `/** @jsxImportSource react */` to the top of the root layout makes `{ compiler: { emotion: true } }` work.

**Comment by dasblitz** [comment]
> You currently need another workaround because of this, as shown in the second @karlhorky's StackBlitz (https://stackblitz.com/edit/vercel-next-js-ajvkxp?file=app/page.tsx) - put `/** @jsxImportSource @emotion/react */` on top of `page.tsx`.

In the linked example, adding the `'use client'` directive to `page.tsx` makes it impossible to use ServerComponents I think? Since they can only be passed as via props or as children, or am I missing something here?

**Comment by Zagorodnyi** [comment]
I've tried everything. This is so cursed

**Comment by tchryssos** [comment]
> Adding `/** @jsxImportSource react */` to the top of the root layout makes `{ compiler: { emotion: true } }` work.

Just want to expand on this / highlight it a little since this lowers the overhead quite a bit for people who DON'T want to disable `compile: { emotion: true }` (and/or the equivalent in `tsconfig` since that seems to throw the same issue)

The vast majority of my files are going to be components that use `styled` or `@mui`, so having my jsx source be `emotion` is pretty important.

As some commenters in this thread and the "workaround" thread have mentioned, you can specify `jsxImportSource` on a file-by-file basis, though most people seem to be doing that to enable emotion specifically on their `'use client'` files; by not specifying the emotion compiler in `next.config.js` you're falling back to regular react for jsx which _does_ work with server components and avoids all the errors people are posting above.

What @Systemcluster posted helped me realize that the opposite was true as well; I could leave my compiler settings intact so that my ~99% of files that are using MUI or some custom styled component only need `'use client'` (which was always going to be the case in `app` directory world, at least for now) since they'll be using the emotion compiler by default, but if I have a server component I can just specify `/** @jsxImportSource react */` explicitly in that file, which will override the compile setting. 

Following other advice here about weaving the client and server components together means that everything works the way I expected after just reading the next docs: I get root layouts that are server (with `metadata` support, which was _actually_ why I was stumbling) and my client children play nicely with them, with "minimum" file prefixing.

**Comment by karlhorky** [comment]
Updates from the React Server Components Emotion ecosystem:

1. Seems like Chakra UI (another UI library dependent on Emotion) is going to switch away from Emotion and to a zero runtime CSS-in-JS solution called Panda:

> ### Zero runtime CSS-in-JS [Panda]
>
> This is the most common and most challenging request we get from users.
>
> Runtime CSS-in-JS and style props are powerful features that allow developers to build dynamic UI components that are composable, predictable, and easy to use. However, it comes at the cost of performance and runtime.
>
> With the release of React Server Components, providing the ability to write Chakra UI components on the server has become crucial. This is a huge win for performance, development, and user experience.
> 
> We're building a new, framework-agnostic styling solution that keeps most of the great features of Chakra's styling system but extracts styles at build time. It'll also feature a PostCSS plugin that extracts styles at postcss run time during development.
>
> Panda will leverage new modern platform features like CSS variables, cascade layers, and W3C token spec.

https://www.adebayosegun.com/blog/the-future-of-chakra-ui

2. MUI is also considering switching to a different solution: https://github.com/mui/material-ui/issues/34826

**Comment by mattrltrent** [comment]
I had to add the `"use client"'` directive to the top of my `page.js` file in `/app`.

**Comment by dzcpy** [comment]
Is there any solution yet?

**Comment by Systemcluster** [comment]
> Is there any solution yet?

Add `/** @jsxImportSource react */` to your layouts, and `"use client"` to your pages and components.

**Comment by karlhorky** [comment]
Thanks for the fix in #54284 @huozhi 🙌 

Once this is released in the next canary version, can someone from this thread who uses Emotion check the new version and provide some proof that it's working in your app? Make sure to install [`next@13.4.20-canary.0`](https://github.com/vercel/next.js/releases/tag/v13.4.20-canary.0) once it's released.

cc @Zagorodnyi @dzcpy @dan-turner @Systemcluster @magnuen2k if you could test as mentioned above, that would be great!

**Comment by dan-turner** [comment]
Will give it a look first thing tomorrow 🙏. 

Love this community ❤️

**Comment by mwskwong** [comment]
> Thanks for the fix in #54284 @huozhi 🙌
> 
> Once this is released in the next canary version, can someone from this thread who uses Emotion check the new version and provide some proof that it's working in your app? Make sure to install [`next@13.4.20-canary.0`](https://github.com/vercel/next.js/releases/tag/v13.4.20-canary.0) (once this version is released).
> 
> cc @Zagorodnyi @dzcpy @dan-turner @Systemcluster @magnuen2k if you could test as mentioned above, that would be great!

Seems to be working for me with MUI:

Repo: https://github.com/mwskwong/resume/commit/f4092af87989e6c56b4bd20e9ad518d482a060ec
Live build: https://resume-82bwpwy6q-mwskwong.vercel.app
